qiskit_experiments/data_processing/data_processor.py

```diff
@@ -163,8 +165,50 @@ def train(self, data: List[Dict[str, Any]]):
             if isinstance(node, TrainableDataAction):
                 if not node.is_trained:
                     # Process the data up to the untrained node.
-                    train_data = []
-                    for datum in data:
-                        train_data.append(self._call_internal(datum, call_up_to_node=index)[0])
+                    node.train(self._call_internal(data, call_up_to_node=index)[0])
+
+    def _data_extraction(self, data: Union[Dict, List[Dict]]) -> List:
+        """Extracts the data on which to run the nodes.
+
+        If the datum is a list of dicts then the data under self._input_key is extracted
+        from each dict and appended to a list which therefore contains all the data. If the
+        data processor has to_array set to True then the list will be converted to a numpy
+        array.
+
+        Args:
+            data: A list of such dicts where the data is contained under the key self._input_key.
+
+        Returns:
+            The data formatted in such a way that it is ready to be processed by the nodes.
+
+        Raises:
+            DataProcessorError:
+                - If the input datum is not a list or a dict.
+                - If the data processor received a single datum but requires all the data to
+                  process it properly.
+                - If the input key of the data processor is not contained in the data.
+        """
+        if isinstance(data, dict):
+            data = [data]
+
+        if isinstance(data, list):
+            try:
+                data_ = [_datum[self._input_key] for _datum in data]
+            except KeyError as error:
+                raise DataProcessorError(
+                    f"The input key {self._input_key} was not found in the input datum."
+                ) from error
+
+        else:
+            raise DataProcessorError(
+                f"{self.__class__.__name__} only extracts data from "
+                f"lists or dicts, received {type(data)}."
+            )
```

Comment:
this is more general, i.e. data might be object type ndarray or some other iterator
```suggestion
        try:
            data_ = [_datum[self._input_key] for _datum in iter(data)]
        except KeyError as error:
            raise DataProcessorError(
                f"The input key {self._input_key} was not found in the input datum."
            ) from error
        except TypeError:
            raise DataProcessorError(
                f"{self.__class__.__name__} only extracts data from "
                f"lists or dicts, received {type(data)}."
            )
```